physical/file/file.go

```diff
@@ -208,20 +221,46 @@ func (b *FileBackend) PutInternal(ctx context.Context, entry *physical.Entry) er
 	}
 
 	// JSON encode the entry and write it
+	fullPath := filepath.Join(path, key)
 	f, err := os.OpenFile(
-		filepath.Join(path, key),
+		fullPath,
 		os.O_CREATE|os.O_TRUNC|os.O_WRONLY,
 		0600)
-	if f != nil {
-		defer f.Close()
-	}
 	if err != nil {
+		if f != nil {
+			f.Close()
+		}
 		return err
 	}
+	if f == nil {
+		return errors.New("could not successfully get a file handle")
+	}
+
 	enc := json.NewEncoder(f)
-	return enc.Encode(&fileEntry{
+	encErr := enc.Encode(&fileEntry{
 		Value: entry.Value,
 	})
+	f.Close()
+	if encErr == nil {
+		return nil
+	}
+
+	// Everything below is best-effort and will result in encErr being returned
+
+	// See if we ended up with a zero-byte file and if so delete it, might be a
+	// case of disk being full but the file info is in metadata that is
+	// reserved.
+	fi, err := os.Stat(fullPath)
+	if err != nil {
+		return encErr
```

Comment:
Since by this point we're in an questionable FS state, it may be useful to wrap/return both `encErr` and `err`.

Reply:
The reason I didn't do that is I wasn't sure it was really meaningful to the user and might just confuse them if they don't realize that we've tried to do some things not specifically related to the original request. E.g. they do a write and get an error back saying that deletion failed.